sycl/source/device_selector.cpp

```diff
@@ -50,36 +52,50 @@ static int getDevicePreference(const device &Device) {
   return Score;
 }
 
-device device_selector::select_device() const {
-  std::vector<device> devices = device::get_devices();
-  int score = REJECT_DEVICE_SCORE;
+static void traceDeviceSelection(const device &Device, int Score, bool Chosen) {
+  bool shouldTrace = false;
+  if (Chosen) {
+    shouldTrace = detail::pi::trace(detail::pi::TraceLevel::PI_TRACE_BASIC);
+  } else {
+    shouldTrace = detail::pi::trace(detail::pi::TraceLevel::PI_TRACE_ALL);
+  }
+  if (shouldTrace) {
+    std::string PlatformName = Device.get_info<info::device::platform>()
+                                   .get_info<info::platform::name>();
+    std::string DeviceName = Device.get_info<info::device::name>();
+    auto selectionMsg = Chosen ? "Selected device: -> final score = "
+                               : "Candidate device: -> score = ";
+
+    std::cout << "SYCL_PI_TRACE[all]: " << selectionMsg << Score
+              << ((Score < 0) ? " (REJECTED)" : "") << std::endl
+              << "SYCL_PI_TRACE[all]: "
+              << "  platform: " << PlatformName << std::endl
+              << "SYCL_PI_TRACE[all]: "
+              << "  device: " << DeviceName << std::endl;
+  }
+}
+
+device select_device(DSelectorInvocableType DeviceSelectorInvocable,
```

Comment:
I guess all these `DSelectorInvocableType` should be `const DSelectorInvocableType&` in all the function/constructor definition to follow the specification, otherwise you might have some copy in the case of lambdas for example.

Reply:
good catch. I can't believe I overlooked that.